src/Microsoft.Diagnostics.Monitoring.StartupHook/Exceptions/Pipeline/Steps/FilterRepeatExceptionPipelineStep.cs

```diff
@@ -0,0 +1,48 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+
+namespace Microsoft.Diagnostics.Monitoring.StartupHook.Exceptions.Pipeline.Steps
+{
+    /// <summary>
+    /// An exception pipeline step that will only allow an exception to be
+    /// processed further if it was not observed previously.
+    /// </summary>
+    internal sealed class FilterRepeatExceptionPipelineStep
+    {
+        // This is not static so that instances of this action can be used
+        // in multiple pipelines that may process the same exception.
+        private readonly object _firstOccurrenceKey = new();
+        private readonly ExceptionPipelineDelegate _next;
+
+        public FilterRepeatExceptionPipelineStep(ExceptionPipelineDelegate next)
+        {
+            ArgumentNullException.ThrowIfNull(next);
+
+            _next = next;
+        }
+
+        public void Invoke(Exception exception)
```

Comment:
While I am not opposed to a loose convention-based contract, if we expect a lot of commonality between Steps (such as a base ctor and a similiar invocation) it might make sense to have a common base class for each step. It depends how guided and flexible you want the next step implementation to be.

Reply:
I leave it as-is and keep this in mind as we write the next set of pipeline steps.